### PR TITLE
Add examples for KVP search

### PR DESCRIPTION
**Pull request by pwalczysko**
Adding examples of KVP search as requested in community thread 
https://github.com/ome/omero-guide-introduction/issues/22

### CODE REVIEW

File: docs/search-omero.rst [hunk 1]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
 
 #. Click on the Browse link |image2|\ in one search result line of the last image (in the right-hand part of the centre pane) to navigate back to the main webclient.
 
+.. note::
+    Spaces, stars, question marks and ``^`` characters are to be avoided in the Keys.
+    If you already have these characters in Keys in your OMERO, 
+    then try to replace them with underscores if you want to 
+    use the Search functionality on them.
+    Spaces stars, question marks and ``^`` characters in Values
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
    Spaces, stars, question marks and ``^`` characters in Values 
```

File: docs/search-omero.rst [hunk 2]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
 
 #. Click on the Browse link |image2|\ in one search result line of the last image (in the right-hand part of the centre pane) to navigate back to the main webclient.
 
+.. note::
+    Spaces, stars, question marks and ``^`` characters are to be avoided in the Keys.
+    If you already have these characters in Keys in your OMERO, 
+    then try to replace them with underscores if you want to 
+    use the Search functionality on them.
+    Spaces stars, question marks and ``^`` characters in Values 
+    are acceptable, but not preferable, see examples below.
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
    are acceptable, but should be avoided if possible, see examples below.
```

File: docs/search-omero.rst [hunk 3]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+
+Key-Value pairs search examples
+-------------------------------
+
+Considering following setup of 12 separate images (image 1 through 12):
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
Considering the following setup of 12 separate images (image 1 through 12):
```

File: docs/search-omero.rst [hunk 4]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+6. Key ``GFP-H2B`` with Value ``2 uM`` on image 6
+7. Key ``GFP_H2B`` with Value ``2_uM`` on image 7
+8. Key ``GFP^H2B`` with Value ``2^uM`` on image 8
+9. Image 9 named ``GFP`` with no Key-Value pairs
+10. Image 10 named ``uM`` with no Key-Value pairs
+11. Image 11 named ``H2B`` with no Key-Value pairs
+12. Image 12 named ``2`` with no Key-Value pairs
+
+Basic **Search** tab:
+   - ``GFP H2B:2 uM`` finds images 1,2,3,5,6,7,8,9,10. This is because the query is interpreted as ``GFP`` OR ``H2B:2`` OR ``uM``.
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
   - ``GFP H2B:2 uM`` finds images 1,2,3,5,6,7,8,9,10. In that case, the query is interpreted as ``GFP`` OR ``H2B:2`` OR ``uM``.
```

File: docs/search-omero.rst [hunk 5]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - ``"GFP H2B":2 uM`` throws an error. Do not use quotes around Key names!
+   - ``GFP H2B:"2 uM"`` finds images 1,2,5,6,7,8,9. This is because the query is interpreted as ``GFP`` OR ``H2B:2 uM`` which prevents finding of image 3 with Value ``2``.
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
   - ``GFP H2B:"2 uM"`` finds images 1,2,5,6,7,8,9. In that case, the query is interpreted as ``GFP`` OR ``H2B:2 uM`` which prevents finding of image 3 with Value ``2``.
```

File: docs/search-omero.rst [hunk 6]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - ``GFP^H2B:2 uM`` finds images 1,2,3,5,6,7,8,9,10. This is because the query is interpreted as ``GFP`` OR ``H2B:2`` OR ``uM``.
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
   - ``GFP^H2B:2 uM`` finds images 1,2,3,5,6,7,8,9,10. In that case, the query is interpreted as ``GFP`` OR ``H2B:2`` OR ``uM``.
```

File: docs/search-omero.rst [hunk 7]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - ``GFP-H2B:"2-uM"`` finds images 5,6.
+   - ``GFP-H2B:"2 uM"`` finds images 5,6.
+   - ``GFP-H2B:"2_uM"`` finds images 5,6.
+   - ``GFP_H2B:2_uM`` finds image 7.
+   - ``GFP^H2B:2^uM`` finds images 1,2,3,5,6,7,8,9,10.
+   - ``GFP`` finds images 1,2,5,6,7,8,9.
+   - ``GFP`` with checkbox ``Name`` under ``Restricted by Field`` section checked finds image 9.
+
+**Advanced** tab:
+   - ``GFP^H2B:2^uM`` and ``GFP^H2B:2 uM`` throw an error in ``Advanced`` tab. This is because of the different interpretation of the ``^`` character between the basic ``Search`` and ``Advanced`` tabs.
```
Comment (jburel) [inline docs/search-omero.rst]:
```suggestion
   - ``GFP^H2B:2^uM`` and ``GFP^H2B:2 uM`` throw an error in ``Advanced`` tab. This is due to the different interpretation of the ``^`` character between the basic ``Search`` and ``Advanced`` tabs.
```

File: docs/search-omero.rst [hunk 8]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - As there is no ``Name`` checkbox in the ``Advanced`` tab, use ``name:GFP`` instead, which finds image 9.
+
+The behaviour for the rest of the query examples in ``Advanced`` tab is the same as listed above for the basic ``Search`` tab.
+
```
Comment (will-moore) [inline docs/search-omero.rst:175]: From this, it sounds like the `Advanced` tab is the same as the regular search, except that it has a couple of limitations. I can see from the walk-through above that you can do `AND..` etc searches with the Advanced search, but I wonder if it's worth describing the difference between them or showing some more examples of what the Advanced search offers? Not sure if it's possible to illustrate that with the Images listed above?

File: docs/search-omero.rst [hunk 9]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
 
 #. Click on the Browse link |image2|\ in one search result line of the last image (in the right-hand part of the centre pane) to navigate back to the main webclient.
 
+.. note::
+    Spaces, stars, question marks and ``^`` characters are to be avoided in the Keys.
+    If you already have these characters in Keys in your OMERO,
```
Comment (will-moore) [inline docs/search-omero.rst]: "OMERO server",

File: docs/search-omero.rst [hunk 10]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+    use the Search functionality on them.
+    Spaces, stars, question marks and ``^`` characters in Values 
+    are acceptable, but should be avoided if possible, see examples below.
+
+Key-Value pairs search examples
+-------------------------------
+
+Considering the following setup of 12 separate images (image 1 through 12):
+
+1. Key ``GFP H2B`` with Value ``2 uM`` on image 1
```
Comment (joshmoore) [inline docs/search-omero.rst]: This might be more readable in a table of: image id, name, key, value

File: docs/search-omero.rst [hunk 11]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+7. Key ``GFP_H2B`` with Value ``2_uM`` on image 7
+8. Key ``GFP^H2B`` with Value ``2^uM`` on image 8
+9. Image 9 named ``GFP`` with no Key-Value pairs
+10. Image 10 named ``uM`` with no Key-Value pairs
+11. Image 11 named ``H2B`` with no Key-Value pairs
+12. Image 12 named ``2`` with no Key-Value pairs
+
+Basic **Search** tab:
+   - ``GFP H2B:2 uM`` finds images 1,2,3,5,6,7,8,9,10. In that case, the query is interpreted as ``GFP`` OR ``H2B:2`` OR ``uM``.
+   - ``"GFP H2B":2 uM`` throws an error. Do not use quotes around Key names!
```
Comment (joshmoore) [inline docs/search-omero.rst]: That this throws an error seems like quite the bug/limitation.

File: docs/search-omero.rst [hunk 12]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - ``GFP H2B:"2 uM"`` finds images 1,2,5,6,7,8,9. In that case, the query is interpreted as ``GFP`` OR ``H2B:2 uM`` which prevents finding of image 3 with Value ``2``.
```
Comment (joshmoore) [inline docs/search-omero.rst:148]: Can any wildecards be added to key values?

File: docs/search-omero.rst [hunk 13]
```diff
@@ -66,6 +66,55 @@ Step-by-Step
...
+   - ``GFP-H2B:"2-uM"`` finds images 5,6.
+   - ``GFP-H2B:"2 uM"`` finds images 5,6.
+   - ``GFP-H2B:"2_uM"`` finds images 5,6.
+   - ``GFP_H2B:2_uM`` finds image 7.
+   - ``GFP^H2B:2^uM`` finds images 1,2,3,5,6,7,8,9,10.
+   - ``GFP`` finds images 1,2,5,6,7,8,9.
+   - ``GFP`` with checkbox ``Name`` under ``Restricted by Field`` section checked finds image 9.
+
+**Advanced** tab:
+   - ``GFP^H2B:2^uM`` and ``GFP^H2B:2 uM`` throw an error in ``Advanced`` tab. This is due to the different interpretation of the ``^`` character between the basic ``Search`` and ``Advanced`` tabs.
```
Comment (joshmoore) [inline docs/search-omero.rst:169]: Again, :+1: for capturing the current state, but this feels like it minimally needs a try/catch around it.

File: docs/search-omero.rst (hunk 8 above)
Comment (pwalczysko) [reply docs/search-omero.rst:175]: > ...what the Advanced search offers?

@will-moore What larger topics would you like to see covered ? The logical operators are covered, in my mind, this is the main strength of the Advanced tab. Do you maybe mean fuzzy search ? What else ?

File: docs/search-omero.rst (hunk 12 above)
Comment (pwalczysko) [reply docs/search-omero.rst:148]: > Can any wildecards be added to key values?

Splitting the question. 
``Can any wildecards be added to keys?`` No. It throws the error https://github.com/ome/omero-guide-introduction/pull/23#discussion_r879399724
``Can any wildecards be added to Values?`` Yes, works intuitively. I will add an example.

File: docs/search-omero.rst (hunk 11 above)
Comment (pwalczysko) [reply docs/search-omero.rst]: Indeed. To explain, this error is "caught" and the user is confronted with a message

`` There was an error parsing your query. Colons ':' are reserved for searches of key-value annotations in the form: 'key:value'. ``

This behaviour is not only limiting, but also the message has a potential to exasperate the user. The user was using the colon in the right way, and this was their main purpose. The error message is highlighting the wrong culprit, and thus not hinting to the solution (the culprit here are the quotes, the colon should stay).
cc @will-moore

File: docs/search-omero.rst (hunk 13 above)
Comment (pwalczysko) [reply docs/search-omero.rst:169]: That is the same error as in https://github.com/ome/omero-guide-introduction/pull/23#discussion_r879399724 - so indeed, it is tried and caught, but not much happiness is coming of it.

**Comment by khaledk2** [comment]
for Key-Value pairs search examples, I think it would be good to add a query to return images 11 and 12?

`Considering the following setup of 12 separate images (image 1 through 12):`

I think the number of images is 13?

**Comment by pwalczysko** [comment]
@khaledk2 

> I think the number of images is 13?

Yes, sorry, I just added a new image.

File: docs/search-omero.rst (hunk 8 above)
Comment (pwalczysko) [reply docs/search-omero.rst:175]: fixed in https://github.com/ome/omero-guide-introduction/pull/23/commits/9003cf5a5c92d00db3fc40b30d557f3a1cf0ffa5

File: docs/search-omero.rst (hunk 9 above)
Comment (pwalczysko) [reply docs/search-omero.rst]: fixed in https://github.com/ome/omero-guide-introduction/pull/23/commits/3e59c0df2de3a87681533e944bf1872f9a18b33c

File: docs/search-omero.rst (hunk 10 above)
Comment (pwalczysko) [reply docs/search-omero.rst]: fixed in https://github.com/ome/omero-guide-introduction/pull/23/commits/86941d46d5c7573855222c80d53df5caae1446fe

File: docs/search-omero.rst (hunk 12 above)
Comment (pwalczysko) [reply docs/search-omero.rst:148]: fixed in https://github.com/ome/omero-guide-introduction/pull/23/commits/e59ace6d928b18d326707a5851266b1fc759fd2d

**Comment by pwalczysko** [comment]
@khaledk2 Thank you for the feedback. Comments fixed in https://github.com/ome/omero-guide-introduction/pull/23/commits/86941d46d5c7573855222c80d53df5caae1446fe

File: docs/search-omero.rst (hunk 11 above)
Comment (pwalczysko) [reply docs/search-omero.rst]: @joshmoore created https://github.com/ome/omero-web/issues/373

**Comment by joshmoore** [comment]
https://omero-guides--23.org.readthedocs.build/projects/introduction/en/23/search-omero.html for review

**Comment by joshmoore** [comment]
:+1: from my side for the table. I could imagine eventually going to more screenshots, but I would hold off until we know where we are going search.

**Comment by khaledk2** [comment]
I think it looks good, but as we have a screenshot for the `Search Box`, we maybe should add another one for the `Advanced search`.

**Comment by will-moore** [comment]
LGTM 👍  Thanks

**Comment by pwalczysko** [comment]
@khaledk2 pushed the screenshot as requested, the current status is at https://omero-guides--23.org.readthedocs.build/projects/introduction/en/23/search-omero.html - please check

**Comment by khaledk2** [comment]
It looks good, thanks! 👍

**Comment by pwalczysko** [comment]
Thanks all. Merging.

**Pull request merged**